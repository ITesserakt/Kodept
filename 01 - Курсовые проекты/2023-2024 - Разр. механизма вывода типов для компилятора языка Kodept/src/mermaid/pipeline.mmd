stateDiagram-v2
    Err: Диагностики
    Le: Лексический анализатор
    Sy: Синтаксический анализатор
    Se: Семантический анализатор
    IR: Генератор промежуточного представления
    Ge: Генератор машинного кода

    [*] --> Le: исходный код программы
    Le --> Sy: поток токенов
    Sy --> Se: AST
    Sy --> IR: AST
    Se --> IR
    IR --> Ge: промежуточное представление
    Ge --> [*]: исполняемый файл

    Se --> Err
    Sy --> Err
    Le --> Err

    Err --> [*]: ошибка компиляции
