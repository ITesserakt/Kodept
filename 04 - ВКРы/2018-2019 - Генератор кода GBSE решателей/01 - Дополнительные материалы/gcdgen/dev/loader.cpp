//
// Created by semyon on 21.05.19.
//

#include "loader.h"
