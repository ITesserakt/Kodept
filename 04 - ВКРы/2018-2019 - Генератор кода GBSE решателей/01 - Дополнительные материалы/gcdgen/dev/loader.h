//
// Created by semyon on 20.05.19.
//

#ifndef CODEGEN_PROJ_LOADER_H
#define CODEGEN_PROJ_LOADER_H

#include <anymap.h>
#include <item.h>

typedef std::string LIB_NAME;
typedef std::string RETURN_TYPE;
typedef std::string TEMPLATE_LOADER_STRING;
typedef std::string FUNC_NAME;

class loader: public item{
private:
    LIB_NAME name;
    RETURN_TYPE ret_type;
    TEMPLATE_LOADER_STRING template_string;
    FUNC_NAME func;
public:
    loader(){
        name = "default_lib";
        ret_type = "HMODULE";
        func = "LoadLibrary";
    };
    loader(LIB_NAME libname, RETURN_TYPE return_type, FUNC_NAME func_name){
        name = libname;
        ret_type = return_type;
        func = func_name;
    };
    std::string get_comment() override{
        return "//Загрузка библиотеки " + name;
    }
    std::string serialize() override {
        return ret_type + " lib_" + name + "=" + func + "(L\""+ name + "\");";
    };
};


#endif //CODEGEN_PROJ_LOADER_H
