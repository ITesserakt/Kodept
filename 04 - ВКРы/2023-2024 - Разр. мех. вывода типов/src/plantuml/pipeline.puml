@startuml
'https://plantuml.com/state-diagram

scale 0.9

state Frontend {
    state "Лексический анализ" as Le

    state "Синтаксический анализ" as Sy
}

state "Middle-end" as M {
    state "Семантический анализ" as Se

    state "Генерация промежуточного представления" as IR
}

state Backend {
    state "Генерация машинного кода" as Ge
}

[*] -right-> Le:исходный код
Le --> Sy: поток токенов
Sy --> Se: AST
Sy --> IR: AST
Se --> IR: символьные таблицы
IR --> Ge: промежуточное представление
Ge --> [*]: объектный файл

@enduml