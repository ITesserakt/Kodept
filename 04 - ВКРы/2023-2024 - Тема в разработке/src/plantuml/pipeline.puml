@startuml
'https://plantuml.com/state-diagram

state "Лексический анализ" as Le
Le: исходный код программы
state "Синтаксический анализ" as Sy
Sy: поток токенов
state "Семантический анализ" as Se
Se: абстрактное синтаксическое \n дерево
state "Генерация промежуточного \n представления" as IR
IR: символьные таблицы
state "Генерация машинного кода" as Ge
Ge: промежуточное представление

[*] --> Le
Le --> Sy
Sy --> Se
Sy --> IR
Se --> IR
IR --> Ge
Ge --> [*]

@enduml