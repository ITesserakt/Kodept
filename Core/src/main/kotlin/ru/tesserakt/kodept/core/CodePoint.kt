package ru.tesserakt.kodept.core

data class CodePoint(val line: Int, val position: Int, val length: Int = 1) {
    override fun toString(): String = "$line:$position"
}