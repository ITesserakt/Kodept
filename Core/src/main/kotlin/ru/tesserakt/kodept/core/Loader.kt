package ru.tesserakt.kodept.core

import java.io.File
import java.nio.file.Path
import java.time.Instant
import kotlin.io.path.Path
import kotlin.io.path.absolute
import kotlin.io.path.extension
import kotlin.io.path.isDirectory

interface Loader {
    fun getSources(): Sequence<CodeSource>

    fun loadSources() = getSources().map { it.contents }
}

class FileLoader private constructor(private val files: Sequence<File>, caches: Sequence<File>) : Loader {
    override fun getSources(): Sequence<CodeSource> = files.map(::FileCodeSource)

    class BuilderScope {
        var path: Path = Path("").absolute()
        var extension = "kd"
        var anySourceExtension = false
        var cacheExtension = "kdc"

        private var cacheFolder: Path? = null

        fun withCaches(scope: Path) {
            cacheFolder = scope
        }

        fun build(): FileLoader {
            require(cacheFolder?.absolute()?.isDirectory() ?: true) { "Provided cache path should be directory" }
            val caches = cacheFolder?.toFile()
                ?.walkTopDown().orEmpty()
                .filter { it.isFile }
                .filter { it.extension == cacheExtension }

            val files = if (path.isDirectory()) {
                require(path.isAbsolute) { "Provided path should be absolute" }
                path.toFile()
                    .walkTopDown()
                    .filter { it.isFile }
                    .filter { anySourceExtension || it.extension == extension }
            } else if (path.toFile().isFile && (anySourceExtension || path.extension == extension)) {
                sequenceOf(path.toFile())
            } else {
                emptySequence()
            }

            return FileLoader(files, caches)
        }
    }

    companion object {
        inline operator fun invoke(scope: BuilderScope.() -> Unit = {}) = BuilderScope().apply(scope).build()
    }
}

class MemoryLoader private constructor(private val streams: Iterable<String>) : Loader {
    // workaround to eager time computation
    override fun getSources(): Sequence<CodeSource> = streams.mapIndexed { i, it ->
        "scratch-#$i-${Instant.now().epochSecond}.kd" to it
    }.asSequence().map { (name, text) ->
        MemoryCodeSource(text.byteInputStream(), name)
    }

    companion object {
        fun fromText(text: Iterable<String>) = MemoryLoader(text)

        fun fromText(vararg text: String) = MemoryLoader(text.toList())

        fun singleSnippet(text: String) = fromText(arrayListOf(text))
    }
}

class PreCachedLoader(private val caches: Sequence<CacheData>) : Loader {
    override fun getSources(): Sequence<CodeSource> = caches.map(::CachedCodeSource)
}