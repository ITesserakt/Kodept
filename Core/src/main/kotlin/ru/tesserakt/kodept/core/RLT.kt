package ru.tesserakt.kodept.core

import arrow.core.Eval
import arrow.core.NonEmptyList

/**
 * Raw lexem tree - it has all information about tokenized lexems
 */
data class RLT(val root: File) {
    sealed interface Node {
        val description: String

        val position: CodePoint
        val text: Eval<String>
    }

    data class Keyword(override val text: Eval<String>, override val position: CodePoint) : Node {
        override val description = text.value()
    }

    sealed class UserSymbol : Node {
        class Identifier(override val text: Eval<String>, override val position: CodePoint) : UserSymbol() {
            override val description = "identifier"

            override fun equals(other: Any?) = other is Identifier && text.value() == other.text.value()
            override fun hashCode(): Int = text.value().hashCode()
        }

        class Type(override val text: Eval<String>, override val position: CodePoint) : UserSymbol() {
            override val description = "type"

            override fun equals(other: Any?) = other is Type && text.value() == other.text.value()
            override fun hashCode(): Int = text.value().hashCode()
        }
    }

    data class Symbol(override val text: Eval<String>, override val position: CodePoint, val type: String) : Node {
        override val description = text.value()
    }

    class ParameterTuple(
        val lparen: Symbol,
        val params: List<Parameter>,
        val rparen: Symbol,
    ) : ExpressionNode, Node by lparen

    open class MaybeTypedParameterTuple(
        val lparen: Symbol,
        open val params: List<MaybeTypedParameter>,
        val rparen: Symbol,
    )

    class TypedParameterTuple(lparen: Symbol, override val params: List<TypedParameter>, rparen: Symbol) :
        MaybeTypedParameterTuple(lparen, params, rparen)

    class Parameter(val id: ExpressionNode) : ExpressionNode, Node by id
    open class MaybeTypedParameter(override val id: UserSymbol.Identifier, open val type: TypeNode?) : Named, Node by id

    class TypedParameter(id: UserSymbol.Identifier, override val type: TypeNode) : MaybeTypedParameter(id, type)

    data class File(val moduleList: NonEmptyList<Module>) : Node by moduleList.head {
        override val description = "file"
    }

    sealed interface TopLevelNode : Node
    sealed interface ObjectLevelNode : Node
    sealed interface StructLevelNode : ObjectLevelNode
    sealed interface TraitLevelNode : ObjectLevelNode
    sealed interface BlockLevelNode : Node
    sealed interface ExpressionNode : BlockLevelNode
    sealed interface StatementNode : BlockLevelNode
    sealed interface Lvalue : Node
    sealed interface TermNode : ExpressionNode, Lvalue
    sealed interface TypeNode : Node
    sealed interface Named : Node {
        val id: UserSymbol
    }

    sealed class Module(val keyword: Keyword, override val id: UserSymbol.Type, val rest: List<TopLevelNode>) : Named,
        Node by id {
        class Global(keyword: Keyword, id: UserSymbol.Type, val flow: Symbol, rest: List<TopLevelNode>) :
            Module(keyword, id, rest)

        class Ordinary(
            keyword: Keyword,
            id: UserSymbol.Type,
            val lbrace: Symbol,
            rest: List<TopLevelNode>,
            val rbrace: Symbol,
        ) : Module(keyword, id, rest)
    }

    data class ForeignType(
        val fKeyword: Keyword,
        val tKeyword: Keyword,
        val id: UserSymbol.Type,
        val flow: Symbol,
        val type: Literal.Text,
    ) : TopLevelNode, Node by id

    data class Struct(
        val keyword: Keyword,
        override val id: UserSymbol.Type,
        val lparen: Symbol?,
        val varsToAlloc: List<TypedParameter>,
        val rparen: Symbol?,
        val lbrace: Symbol?,
        val rest: List<StructLevelNode>,
        val rbrace: Symbol?,
    ) : TopLevelNode, Named, Node by id

    data class Trait(
        val keyword: Keyword,
        override val id: UserSymbol.Type,
        val lbrace: Symbol?,
        val rest: List<TraitLevelNode>,
        val rbrace: Symbol?,
    ) : TopLevelNode, Named, Node by id

    sealed class Enum(
        val keyword: Keyword,
        override val id: UserSymbol.Type,
        val lbrace: Symbol?,
        val rest: NonEmptyList<UserSymbol.Type>,
        val rbrace: Symbol?,
    ) : TopLevelNode, Named, Node by id {
        class Stack(
            keyword: Keyword,
            id: UserSymbol.Type,
            lbrace: Symbol?,
            rest: NonEmptyList<UserSymbol.Type>,
            rbrace: Symbol?,
        ) : Enum(keyword, id, lbrace, rest, rbrace)

        class Heap(
            keyword: Keyword,
            id: UserSymbol.Type,
            lbrace: Symbol?,
            rest: NonEmptyList<UserSymbol.Type>,
            rbrace: Symbol?,
        ) : Enum(keyword, id, lbrace, rest, rbrace)
    }

    sealed interface Body : ExpressionNode {
        data class Expression(val flow: Symbol, val expression: ExpressionNode) : Body, ExpressionNode by expression
        data class Block(val lbrace: Symbol, val block: List<BlockLevelNode>, val rbrace: Symbol) : Body, Node by lbrace
    }

    sealed class Function(
        val keyword: Keyword,
        override val id: UserSymbol.Identifier,
        open val params: List<MaybeTypedParameterTuple>,
        val colon: Symbol?,
        open val returnType: TypeNode?,
    ) : Named, Node by id {
        class Abstract(
            keyword: Keyword,
            id: UserSymbol.Identifier,
            override val params: List<TypedParameterTuple>,
            colon: Symbol?,
            returnType: TypeNode?,
        ) : Function(keyword, id, params, colon, returnType), TraitLevelNode

        class Bodied(
            keyword: Keyword,
            id: UserSymbol.Identifier,
            params: List<MaybeTypedParameterTuple>,
            colon: Symbol?,
            returnType: TypeNode?,
            val body: Body,
        ) : Function(keyword, id, params, colon, returnType), TopLevelNode, StructLevelNode, StatementNode,
            TraitLevelNode

        class Foreign(
            keyword: Keyword,
            id: UserSymbol.Identifier,
            override val params: List<TypedParameterTuple>,
            colon: Symbol?,
            override val returnType: Reference?,
            val descriptor: Literal.Text,
        ) : Function(keyword, id, params, colon, returnType), TopLevelNode
    }

    open class Reference(val ref: UserSymbol) : TermNode, TypeNode, Node by ref

    data class Application(val expr: ExpressionNode, val params: List<Parameter>) : ExpressionNode, Node by expr

    sealed interface Context {
        object StartsFromRoot

        val global: Boolean

        data class Global(val colon: Symbol) : Context {
            override val global = true

            override fun unfold() = StartsFromRoot to emptyList<Reference>()
        }

        object Local : Context {
            override val global = false

            override fun unfold() = null to emptyList<Reference>()
        }

        data class Inner(val type: Reference, val parent: Context) : Context {
            override val global = parent.global

            override fun unfold() = parent.unfold().let { it.first to it.second + type }
        }

        fun unfold(): Pair<StartsFromRoot?, List<Reference>>
    }

    class ContextualReference(val context: Context, reference: Reference) : Reference(reference.ref)

    sealed class Variable(
        val keyword: Keyword,
        override val id: UserSymbol.Identifier,
        val colon: Symbol?,
        val type: TypeNode?,
    ) : StatementNode, Lvalue, Named, Node by id {
        class Immutable(keyword: Keyword, id: UserSymbol.Identifier, colon: Symbol?, type: TypeNode?) :
            Variable(keyword, id, colon, type)

        class Mutable(keyword: Keyword, id: UserSymbol.Identifier, colon: Symbol?, type: TypeNode?) :
            Variable(keyword, id, colon, type)
    }

    open class Assignment(open val lvalue: Lvalue, val equals: Symbol, val expression: ExpressionNode) : StatementNode,
        Node by equals

    class InitializedAssignment(override val lvalue: Variable, equals: Symbol, expression: ExpressionNode) :
        Assignment(lvalue, equals, expression), Named {
        override val id: UserSymbol = lvalue.id
    }

    class CompoundAssignment(lvalue: Lvalue, val compoundOperator: Symbol, expression: ExpressionNode) :
        Assignment(lvalue, compoundOperator, expression)

    data class BinaryOperation(val left: ExpressionNode, val op: Symbol, val right: ExpressionNode) : ExpressionNode,
        Node by op

    data class Access(val left: ExpressionNode, val dot: Symbol, val right: ExpressionNode) : ExpressionNode,
        Node by dot

    data class UnaryOperation(val expression: ExpressionNode, val op: Symbol) : ExpressionNode, Node by op

    sealed class Literal(override val text: Eval<String>, override val position: CodePoint) : ExpressionNode {
        override val description get() = text.value()

        class Number(text: Eval<String>, position: CodePoint) : Literal(text, position)
        class Floating(text: Eval<String>, position: CodePoint) : Literal(text, position)
        class Text(text: Eval<String>, position: CodePoint) : Literal(text, position) {
            fun isChar() = text.value().first() == '\''
            fun isString() = text.value().first() == '"'
        }

        data class Tuple(val lparen: Symbol, val expressions: List<ExpressionNode>, val rparen: Symbol) :
            Literal(lparen.text, lparen.position)
    }

    data class If(
        val keyword: Keyword,
        val condition: ExpressionNode,
        val body: Body,
        val elifs: List<Elif>,
        val el: Else?,
    ) : ExpressionNode, Node by keyword {
        data class Elif(val keyword: Keyword, val condition: ExpressionNode, val body: Body) : Node by keyword
        data class Else(val keyword: Keyword, val body: Body) : Node by keyword
    }

    data class While(
        val keyword: Keyword,
        val condition: ExpressionNode,
        val body: Body,
    ) : StatementNode, Node by keyword

    data class Match(val match: Keyword, val receiver: ExpressionNode?, val branches: NonEmptyList<Branch>) :
        ExpressionNode, Node by match {
        data class Branch(val condition: ExpressionNode, val body: Body.Expression) : Node by condition
    }

    data class TupleType(val lparen: Symbol, val types: List<TypeNode>, val rparen: Symbol) : TypeNode, Node by lparen

    data class UnionType(val lparen: Symbol, val types: NonEmptyList<TypeNode>, val rparen: Symbol) : TypeNode,
        Node by lparen

    data class Lambda(val params: List<Reference>, val flow: Symbol, val body: ExpressionNode) : ExpressionNode,
        Node by flow

    data class Extension(
        val extend: Keyword,
        val onType: Reference,
        val with: Keyword,
        val forTrait: Reference,
        val body: List<Function.Bodied>,
    ) : TopLevelNode, Node by extend
}