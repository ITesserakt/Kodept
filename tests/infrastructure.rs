use std::borrow::Cow;
use std::fmt::Display;
use std::path::Path;

use insta::assert_debug_snapshot;
use nom_supreme::final_parser::final_parser;

use kodept::loader::Loader;
use kodept::macro_context::DefaultContext;
use kodept::parse_error::Reportable;
use kodept::read_code_source::ReadCodeSource;
use kodept::steps::hlist::{HCons, HNil};
use kodept::steps::pipeline::Pipeline;
use kodept::steps::Step;
use kodept::top_parser;
use kodept_ast::ast_builder::ASTBuilder;
use kodept_core::code_point::CodePoint;
use kodept_core::code_source::CodeSource;
use kodept_core::structure::rlt::RLT;
use kodept_core::structure::span::CodeHolder;
use kodept_inference::assumption::Assumptions;
use kodept_interpret::operator_desugaring::*;
use kodept_interpret::semantic_analyzer::ScopeAnalyzer;
use kodept_interpret::type_checker::TypeChecker;
use kodept_interpret::Witness;
use kodept_macros::error::report_collector::ReportCollector;
use kodept_macros::traits::{MutableContext, UnrecoverableError};
use kodept_parse::ParseError;
use kodept_parse::token_stream::TokenStream;
use kodept_parse::tokenizer::Tokenizer;

pub const EXAMPLES_FOLDER: &str = "examples";

struct CodeProvider<'a>(&'a str);

impl CodeHolder for CodeProvider<'_> {
    fn get_chunk(&self, at: CodePoint) -> Cow<str> {
        Cow::Borrowed(&self.0[at.offset..at.offset + at.length])
    }
}

fn get_code_source(name: impl Display) -> CodeSource {
    let loader = Loader::file()
        .with_starting_path(Path::new(&format!("{EXAMPLES_FOLDER}/{name}.kd")))
        .build()
        .expect("Cannot get given path");

    loader
        .into_sources()
        .pop()
        .expect("No source found with provided name")
}

fn get_rlt(source: &ReadCodeSource) -> RLT {
    let tokens = Tokenizer::new(source.contents()).into_vec();
    let tokens = TokenStream::new(&tokens);
    let result = final_parser(top_parser)(tokens)
        .map_err(|it: ParseError| it.to_diagnostics())
        .expect("Cannot parse");
    result
}

fn common_steps(ctx: &mut impl MutableContext) -> Result<Vec<Assumptions>, UnrecoverableError> {
    Pipeline
        .define_step(HCons {
            head: AccessExpander::new(),
            tail: HCons {
                head: BinaryOperatorExpander::new(),
                tail: HCons {
                    head: UnaryOperatorExpander::new(),
                    tail: HNil,
                },
            },
        })
        .apply_with_context(ctx)?;

    let mut scopes = ScopeAnalyzer::new();
    Pipeline
        .define_step(HCons {
            head: &mut scopes,
            tail: HNil,
        })
        .apply_with_context(ctx)?;
    let scopes = scopes.into_inner();

    let mut type_checker = TypeChecker::new(
        &scopes,
        Witness::fact(
            AccessExpander,
            BinaryOperatorExpander,
            UnaryOperatorExpander,
        ),
    );
    Pipeline
        .define_step(HCons {
            head: &mut type_checker,
            tail: HNil,
        })
        .apply_with_context(ctx)?;
    Ok(type_checker.into_inner())
}

fn test_typing(name: &str) {
    let source = ReadCodeSource::try_from(get_code_source(name)).expect("Cannot read source");
    let provider = CodeProvider(source.contents());
    let rlt = get_rlt(&source);

    let (ast, rlt_accessor) = ASTBuilder.recursive_build(&rlt.0, &provider);
    let ast = ast.build();

    let mut context = DefaultContext::new(
        source.with_filename(|_| ReportCollector::new()),
        rlt_accessor,
        ast,
    );

    assert_debug_snapshot!(common_steps(&mut context).expect("Success"));
}

#[test]
fn test_typing_on_church_encoding() {
    test_typing("church")
}

#[test]
fn test_typing_on_test_file() {
    test_typing("test")
}

#[test]
#[should_panic]
fn test_typing_on_rule110() {
    test_typing("rule110")
}

#[test]
#[should_panic]
fn test_typing_on_fibonacci() {
    test_typing("fibonacci")
}
