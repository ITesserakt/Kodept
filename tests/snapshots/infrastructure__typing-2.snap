---
source: tests/infrastructure.rs
expression: values
---
[]
